x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java

```diff
@@ -267,6 +270,23 @@ private static boolean isCreatedOnLatestVersion(IndexMetadata indexMetadata) {
         return indexVersionCreated != null && indexVersionCreated.onOrAfter(IndexVersion.current());
     }
 
+    /**
+     * Check to see if there are any file based role mappings and if they have been loaded into cluster state. If the
+     * {@link ReservedStateMetadata} file_settings namespace contains role mapping names, it means that there should be the same number of
+     * cluster state role mappings available. If they're not available yet using {@code RoleMappingMetadata.getFromClusterState()}, they
+     * have not yet been synchronized.
+     */
+    private static boolean isReservedRoleMappingsSynced(ClusterState clusterState) {
```

Comment:
WDYT about calling this `readyForRoleMappingCleanupMigration`? It feels more relevant as a security-index concept. 

Alternatively, I _think_ we could entirely remove this from the index manager if we expose a `client` instance (or the `ClusterService`) to the `checkPreConditions` method. That would give us cluster state instead of relying on the index manager.

Reply:
> WDYT about calling this readyForRoleMappingCleanupMigration

My intention was to decouple it from the migration but I guess that's pretty silly, because it's only ever going to be used for that. I'll update it to be more explicit. 

>Alternatively, I think we could entirely remove this from the index manager if we expose a client instance (or the ClusterService) to the checkPreConditions method. That would give us cluster state instead of relying on the index manager.

The migrations are only triggered on security index manager state changes (it's a listener). So if the migration is not ready to be executed, we need a way to make sure that the state changes when it is and that's why this needs to be part of the state. 

That's what I'm referring to here: https://github.com/elastic/elasticsearch/pull/114830#discussion_r1801226195 ideally I think there should be a separate state that track the migrations, where we can track state changes that are not necessarily specific to the security index.